### PR TITLE
xds: support new fields to fetch security configuration

### PR DESCRIPTION
**Pull request by easwars**
The new method is documented at https://github.com/grpc/proposal/blob/master/A29-xds-tls-security.md and based on envoyproxy/envoy#17201. The old fields are still supported and are used as fallback when the new fields don't provide the required data.

TL;dr
- Identity certs used to be fetched from the `tls_certificate_certificate_provider_instance` field [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L256).
  - Going forward, they will be fetched from the `tls_certificate_provider_instance` field [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L247).
- Validation context used to be fetched from `validation_context_certificate_provider_instance` field [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L283) or from the `validation_context_certificate_provider_instance` field of `CombinedCertificateValidationContext` [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L205)
  - Going forward, they will be fetched from the `ca_certificate_provider_instance` field of `CertificateValidationContext` [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L315)
  - The `CertificateValidationContext` can be fetched either from the `validation_context` field of the `validation_context_type` oneof [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L261) or from the `default_validation_context` field of the `combined_validation_context` [here](https://github.com/envoyproxy/envoy/blob/ff68b337d038acccb770fcceec57441cf0c740e2/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L188).

RELEASE NOTES:
- Support fetching of security configuration from newly added fields in envoyproxy/envoy#17201

### CODE REVIEW

File: xds/internal/xdsclient/xds.go [hunk 1]
```diff
@@ -812,18 +812,39 @@ func securityConfigFromCluster(cluster *v3clusterpb.Cluster) (*SecurityConfig, e
...
-	if err != nil {
-		return nil, err
+	return securityConfigFromCommonTLSContext(upstreamCtx.GetCommonTlsContext(), false)
+}
+
+// common is expected to be not nil.
+func securityConfigFromCommonTLSContext(common *v3tlspb.CommonTlsContext, server bool) (*SecurityConfig, error) {
+	sc := securityConfigFromCommonTLSContextUsingNewFields(common)
+	var err error
+	if sc == nil || sc.Equal(&SecurityConfig{}) {
```
Comment (menghanl) [inline xds/internal/xdsclient/xds.go:822]: Why does the function return `nil` or an empty `SecurityConfig`? Can it just return `nil`?

File: xds/internal/xdsclient/xds.go [hunk 2]
```diff
@@ -873,6 +894,80 @@ func securityConfigFromCommonTLSContext(common *v3tlspb.CommonTlsContext) (*Secu
...
+	//      the bootstrap configuration
+	//  - certificate_name
+	//    -  this is an opaque name passed to the certificate provider
+	var validationCtx *v3tlspb.CertificateValidationContext
+	switch common.GetValidationContextType().(type) {
+	case *v3tlspb.CommonTlsContext_ValidationContext:
+		validationCtx = common.GetValidationContext()
+	case *v3tlspb.CommonTlsContext_CombinedValidationContext:
+		validationCtx = common.GetCombinedValidationContext().GetDefaultValidationContext()
+	case nil:
```
Comment (menghanl) [inline xds/internal/xdsclient/xds.go:945]: You don't need this case, right?
The following `validationCtx == nil` will handle this and return nil.

File: xds/internal/xdsclient/xds.go [hunk 3]
```diff
@@ -873,6 +894,80 @@ func securityConfigFromCommonTLSContext(common *v3tlspb.CommonTlsContext) (*Secu
...
+
+	if rootProvider := validationCtx.GetCaCertificateProviderInstance(); rootProvider != nil {
+		sc.RootInstanceName = rootProvider.GetInstanceName()
+		sc.RootCertName = rootProvider.GetCertificateName()
+	}
+	var matchers []matcher.StringMatcher
+	for _, m := range validationCtx.GetMatchSubjectAltNames() {
+		matcher, err := matcher.StringMatcherFromProto(m)
+		if err != nil {
+			return nil
```
Comment (menghanl) [inline xds/internal/xdsclient/xds.go:963]: What does this error mean? Is it OK to ignore? Should it NACK?

**Review by menghanl — CHANGES_REQUESTED** [review]

File: xds/internal/xdsclient/xds.go (hunk 2 above)
Comment (easwars) [reply xds/internal/xdsclient/xds.go:947]: We need this case when only the `identity` certs are specified on the server side. Returning `sc` here would ensure that whatever identity certs were fetched in lines 911-914 would get returned.

Comment (easwars) [reply xds/internal/xdsclient/xds.go:947]: This is again the same reason why we need to check for empty `SecurityConfig` in line 822, where you also commented.

File: xds/internal/xdsclient/xds.go (hunk 1 above)
Comment (easwars) [reply xds/internal/xdsclient/xds.go:824]: See other comment.

File: xds/internal/xdsclient/xds.go (hunk 3 above)
Comment (easwars) [reply xds/internal/xdsclient/xds.go]: You are right. Thanks. Fixed it.

**Review by menghanl — APPROVED** [review]

**Pull request merged**